File: scripts/build-updating-metadata.py
import csv
import json
import re
import sys
import unicodedata
from collections import defaultdict
from difflib import SequenceMatcher
from pathlib import Path

from openpyxl import load_workbook


ROOT = Path(__file__).resolve().parents[1]
DOWNLOADS = Path(sys.argv[1]) if len(sys.argv) > 1 else Path.home() / "Downloads"


def read_csv(name: str):
    with (DOWNLOADS / name).open(encoding="utf-8-sig", newline="") as handle:
        return list(csv.DictReader(handle))


def normalize(value: str):
    value = unicodedata.normalize("NFKD", value.casefold()).encode("ascii", "ignore").decode()
    value = value.replace("&", " dan ")
    value = re.sub(r"\b(kepala dan anggota keluarga|kepala keluarga|anggota keluarga|responden)\b", " ", value)
    return " ".join(re.sub(r"[^a-z0-9]+", " ", value).split())


def role_name(subject: str):
    if subject == "anggota keluarga":
        return "MEMBER"
    if subject == "kepala keluarga":
        return "HEAD"
    return "ALL"


def frequency_name(value: str):
    return {
        "Insidentil": "INCIDENTAL",
        "6 bulan": "SIX_MONTHS",
        "Tahunan": "ANNUAL",
        "Tidak Berubah": "IMMUTABLE",
    }.get(value)


mapping = json.loads((ROOT / "config" / "indikator-mapping.json").read_text(encoding="utf-8"))["kolom"]
merge_rows = read_csv("merge_pertanyaan.csv")
answer_rows = read_csv("jawabantertutup.csv")
question_rows = read_csv("questions(logic_jawaban).csv")

datatype_workbook = load_workbook(DOWNLOADS / "ajaib_datatype.xlsx", data_only=True)
datatype_sheet = datatype_workbook["ajaib_datatype"]
datatype_status = {}
for table, column, _data_type, status in datatype_sheet.iter_rows(min_row=2, values_only=True):
    if str(table or "").strip().casefold() == "ajaib" and column:
        datatype_status[str(column).strip()] = str(status or "").strip().casefold() or None

NON_EDITABLE_DATATYPE_STATUSES = {"old", "changed", "temporer", "perhitungan sistem"}

workbook = load_workbook(DOWNLOADS / "Parameter_DDP_Estimasi_Waktu_Updating (1).xlsx", data_only=True)
sheet = workbook["Semua Parameter"]
excel_rows = []
for row in sheet.iter_rows(min_row=4, values_only=True):
    if row[3] and frequency_name(str(row[5] or "").strip()):
        excel_rows.append(
            {
                "label": str(row[3]).strip(),
                "normalized": normalize(str(row[3])),
                "frequency": frequency_name(str(row[5]).strip()),
            }
        )

answers = defaultdict(list)
for row in answer_rows:
    if row["status"].strip() != "0":
        continue
    value = row["value"].strip()
    if value and value not in answers[row["pertanyaan_id"]]:
        answers[row["pertanyaan_id"]].append(value)

merge_by_id = {row["id"]: row for row in merge_rows}
questions_by_id = {row["id"]: row for row in question_rows}
merge_by_header = defaultdict(list)
for row in merge_rows:
    header = row["header"].strip()
    if header:
        merge_by_header[header].append(row)

EXPLICIT_LABELS = {
    "kode_deskel": "Kode Desa",
    "kode_bangunan": "Kode Bangunan",
    "deskel": "Desa/Kelurahan",
    "dusun": "Dusun/Kampung/ Dukuh",
    "rw": "RW",
    "rt": "RT",
    "responden": "Nama Responden",
    "kesediaan": "kesediaan sebagai responden",
    "nkk": "Nomor KK",
    "nama": "Nama Kepala Keluarga",
    "nik": "NIK Kepala Keluarga",
    "no_hp": "Nomor HP Responden",
    "dead_jml": "Jumlah Anggota Keluarga yang Meninggal",
    "thn_datang": "Tahun Datang ke Desa",
    "menetap": "Tinggal Menetap",
    "dinamika": "Dinamika Kehidupan",
    "tgl_lahir": "Tanggal Lahir Kepala Keluarga",
    "punya_aktalahir": "Kepemilikan Akta Lahir Kepala Keluarga dan Anggota Keluarga",
    "kendaraan_jml": "Jumlah Kendaraan Bermotor Dalam Rumah",
    "anak_asi": "ASI Eklusif",
    "rp_listrik": "Biaya Listrik Per bulan (Rp)",
    "rumah_kamar": "Jumah Kamar Tidur",
}


def field_frequency(field: str, rows):
    explicit = EXPLICIT_LABELS.get(field)
    if explicit:
        target = normalize(explicit)
        exact = next((row for row in excel_rows if row["normalized"] == target), None)
        if exact:
            return exact["frequency"], exact["label"], 1.0

    candidates = [normalize(row["value"]) for row in rows if row["value"].strip()]
    if not candidates:
        candidates = [normalize(field.replace("_", " "))]
    best = None
    for candidate in candidates:
        for excel in excel_rows:
            score = 1.0 if candidate == excel["normalized"] else SequenceMatcher(None, candidate, excel["normalized"]).ratio()
            if best is None or score > best[0]:
                best = (score, excel)
    if best and best[0] >= 0.90:
        return best[1]["frequency"], best[1]["label"], round(best[0], 3)
    return None, None, round(best[0], 3) if best else 0


fields = {}
unmapped_frequency = []
for field in mapping:
    rows = merge_by_header.get(field, [])
    frequency, source_label, confidence = field_frequency(field, rows)
    if not frequency:
        unmapped_frequency.append(field)
    variants = {}
    for row in rows:
        role = role_name(row["subjek"].strip())
        question = questions_by_id.get(row["id"])
        variant = variants.setdefault(role, {})
        variant["active"] = bool(variant.get("active")) or row["status"].strip() == "0"
        if answers[row["id"]]:
            variant["options"] = answers[row["id"]]
        if question and question["status"].strip() == "0":
            variant["inputType"] = question["question_type"].strip() or None
            description = " ".join(question["description"].split())
            if description:
                variant["help"] = description
            try:
                conditions = json.loads(question["conditions"] or "{}")
            except json.JSONDecodeError:
                conditions = {}
            show_if = conditions.get("show_if") if isinstance(conditions, dict) else None
            if isinstance(show_if, dict):
                source = merge_by_id.get(str(show_if.get("question_id")), {})
                source_field = source.get("header", "").strip()
                if source_field:
                    values = show_if.get("value", [])
                    variant["condition"] = {
                        "field": source_field,
                        "values": values if isinstance(values, list) else [values],
                    }
    status = datatype_status.get(field)
    datatype_editable = status not in NON_EDITABLE_DATATYPE_STATUSES
    fields[field] = {
        "frequency": frequency,
        "frequencySource": source_label,
        "frequencyConfidence": confidence,
        "datatypeStatus": status,
        "editable": datatype_editable and any(bool(variant.get("active")) for variant in variants.values()),
        "variants": variants,
    }

event_labels = {
    "deathCount": "Jumlah Anggota Keluarga yang Meninggal",
    "deathName": "Nama anggota keluarga yang meninggal",
    "deathNik": "NIK Anggota keluarga yang meninggal",
    "deathBirthDate": "Tanggal Lahir anggota keluarga yang meninggal",
    "deathDate": "Tanggal kematian anggota keluarga yang meninggal",
    "deathCause": "Penyebab Kematian anggota keluarga yang meninggal",
    "deathCertificate": "Kepemilikan Akta Kematian anggota keluarga yang meninggal",
}
events = {}
for key, label in event_labels.items():
    normalized = normalize(label)
    excel = next((row for row in excel_rows if row["normalized"] == normalized), None)
    merge = next((row for row in merge_rows if normalize(row["value"]) == normalized), None)
    events[key] = {
        "label": label,
        "frequency": excel["frequency"] if excel else "INCIDENTAL",
        "options": answers[merge["id"]] if merge else [],
    }

output = {
    "_meta": {
        "sources": [
            "Parameter_DDP_Estimasi_Waktu_Updating (1).xlsx",
            "jawabantertutup.csv",
            "merge_pertanyaan.csv",
            "questions(logic_jawaban).csv",
            "ajaib_datatype.xlsx",
        ],
        "tooltipPolicy": "QUESTION_LOGIC_ONLY",
        "unmappedFrequency": unmapped_frequency,
    },
    "fields": fields,
    "events": events,
    "building": {
        "kode": {"frequency": "ANNUAL"},
        "jenis": {"frequency": "ANNUAL"},
        "kategori": {"frequency": "ANNUAL"},
        "keterangan": {"frequency": "ANNUAL"},
        "fotoUrl": {"frequency": "ANNUAL"},
        "dusun": {"frequency": "ANNUAL"},
        "rw": {"frequency": "ANNUAL"},
        "rt": {"frequency": "ANNUAL"},
    },
}

(ROOT / "config" / "updating-metadata.json").write_text(
    json.dumps(output, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
)
print(f"Wrote metadata for {len(fields)} fields; {len(unmapped_frequency)} frequencies remain unspecified.")

File: scripts/build-updating-metadata.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const DEFAULT_SOURCE = path.join(ROOT, "config", "sources", "merge_pertanyaan.csv");
const MAPPING_PATH = path.join(ROOT, "config", "indikator-mapping.json");
const METADATA_PATH = path.join(ROOT, "config", "updating-metadata.json");
const RETIRED_PATH = path.join(ROOT, "config", "kolom-dipensiunkan.json");

const FREQUENCIES = {
  Insidentil: "INCIDENTAL",
  "6 bulan": "SIX_MONTHS",
  Tahunan: "ANNUAL",
  "Tidak Berubah": "IMMUTABLE",
};
const ROLES = {
  "kepala keluarga": "HEAD",
  "anggota keluarga": "MEMBER",
};
const PRIORITY = { IMMUTABLE: 0, INCIDENTAL: 1, ANNUAL: 2, SIX_MONTHS: 3 };
const CONDITION_ONLY = new Set(["nel_kategori", "penyakit_jumlah"]);
const DERIVED = new Set([
  "miskin_bps",
  "miskin_ekstrem",
  "miskin_uufm",
  "miskin_wb",
  "miskin_bpsd",
  "skor_kls",
  "ppkb",
  "ppkt",
  "pkb",
]);

function readJson(file) {
  if (!fs.existsSync(file)) throw new Error(`Required source is missing: ${file}`);
  return JSON.parse(fs.readFileSync(file, "utf8"));
}

function parseDelimited(content) {
  const rows = [];
  let row = [];
  let value = "";
  let quoted = false;
  for (let index = 0; index < content.length; index += 1) {
    const char = content[index];
    if (char === '"') {
      if (quoted && content[index + 1] === '"') {
        value += '"';
        index += 1;
      } else {
        quoted = !quoted;
      }
    } else if (char === ";" && !quoted) {
      row.push(value);
      value = "";
    } else if ((char === "\n" || char === "\r") && !quoted) {
      if (char === "\r" && content[index + 1] === "\n") index += 1;
      row.push(value);
      value = "";
      if (row.some((cell) => cell.length)) rows.push(row);
      row = [];
    } else {
      value += char;
    }
  }
  if (value.length || row.length) {
    row.push(value);
    rows.push(row);
  }
  const headers = rows.shift()?.map((cell) => cell.replace(/^\uFEFF/, "")) ?? [];
  return rows.map((cells) =>
    Object.fromEntries(headers.map((header, index) => [header, cells[index] ?? ""])),
  );
}

function chooseFrequency(byRole) {
  const values = Object.values(byRole);
  return values.length
    ? values.reduce((best, current) => (PRIORITY[current] > PRIORITY[best] ? current : best))
    : null;
}

function main() {
  const args = process.argv.slice(2);
  const importIndex = args.indexOf("--import-source");
  const shouldImport = importIndex >= 0;
  if (shouldImport) args.splice(importIndex, 1);
  let source = path.resolve(args[0] ?? DEFAULT_SOURCE);
  if (fs.existsSync(source) && fs.statSync(source).isDirectory()) {
    source = path.join(source, "merge_pertanyaan.csv");
  }
  if (!fs.existsSync(source)) {
    throw new Error(
      `Question source is missing: ${source}. Pass the CSV path and --import-source on the first run.`,
    );
  }
  if (shouldImport) {
    fs.mkdirSync(path.dirname(DEFAULT_SOURCE), { recursive: true });
    if (source !== path.resolve(DEFAULT_SOURCE)) fs.copyFileSync(source, DEFAULT_SOURCE);
    source = DEFAULT_SOURCE;
  }

  const mapping = readJson(MAPPING_PATH).kolom;
  const previous = readJson(METADATA_PATH);
  const retired = new Set(readJson(RETIRED_PATH).columns);
  const rows = parseDelimited(fs.readFileSync(source, "utf8"));

  const byFieldRole = new Map();
  const labels = new Map();
  const skippedUnknown = new Set();
  const statusCounts = new Map();
  const rawRoleFields = { HEAD: new Set(), MEMBER: new Set() };

  for (const row of rows) {
    const status = (row.status ?? "").trim();
    statusCounts.set(status, (statusCounts.get(status) ?? 0) + 1);
    if (status !== "0") continue;
    const field = (row.header ?? "").trim();
    const role = ROLES[(row.subjek ?? "").trim().toLocaleLowerCase("id-ID")];
    const frequency = FREQUENCIES[(row["periode updating"] ?? "").trim()];
    if (!field || !role || !frequency) continue;
    rawRoleFields[role].add(field);
    if (!(field in mapping)) {
      skippedUnknown.add(field);
      continue;
    }
    const byRole = byFieldRole.get(field) ?? {};
    if (byRole[role] && byRole[role] !== frequency) {
      throw new Error(`Conflicting frequency for ${field}/${role}: ${byRole[role]} vs ${frequency}`);
    }
    byRole[role] = frequency;
    byFieldRole.set(field, byRole);
    if (!labels.has(field)) labels.set(field, (row.value ?? "").trim());
  }

  const fields = {};
  for (const field of Object.keys(mapping)) {
    const old = { ...(previous.fields?.[field] ?? {}) };
    const frequencyByRole = byFieldRole.get(field) ?? {};
    const askedTo = ["HEAD", "MEMBER"].filter((role) => role in frequencyByRole);
    const conditionOnly = CONDITION_ONLY.has(field);
    const deprecated = retired.has(field);
    const variants = { ...(old.variants ?? {}) };
    if (conditionOnly) {
      for (const role of ["HEAD", "MEMBER"]) {
        variants[role] = { ...(variants[role] ?? {}), active: true };
      }
    }
    fields[field] = {
      ...old,
      frequency: chooseFrequency(frequencyByRole),
      frequencyByRole,
      askedTo,
      frequencySource: labels.get(field) ?? null,
      frequencyConfidence: Object.keys(frequencyByRole).length ? 1 : 0,
      editable:
        !deprecated &&
        (conditionOnly || Object.values(variants).some((variant) => Boolean(variant?.active))),
      deprecated,
      conditionOnly,
      derived: DERIVED.has(field),
      variants,
    };
  }

  const head = new Set([...byFieldRole].filter(([, roles]) => "HEAD" in roles).map(([field]) => field));
  const member = new Set([...byFieldRole].filter(([, roles]) => "MEMBER" in roles).map(([field]) => field));
  const periodicByRole = (frequency, role) =>
    new Set(
      [...byFieldRole]
        .filter(([, roles]) => roles[role] === frequency)
        .map(([field]) => field),
    );
  const periodic = (frequency) =>
    new Set(
      [...byFieldRole]
        .filter(([, roles]) => Object.values(roles).includes(frequency))
        .map(([field]) => field),
    );
  const stats = {
    totalRows: rows.length,
    activeRows: statusCounts.get("0") ?? 0,
    inactiveRows: statusCounts.get("1") ?? 0,
    classifiedSchemaFields: byFieldRole.size,
    headFields: rawRoleFields.HEAD.size,
    memberFields: rawRoleFields.MEMBER.size,
    headOnlyFields: [...rawRoleFields.HEAD].filter((field) => !rawRoleFields.MEMBER.has(field)).length,
    memberOnlyFields: [...rawRoleFields.MEMBER].filter((field) => !rawRoleFields.HEAD.has(field)).length,
    sharedFields: [...rawRoleFields.HEAD].filter((field) => rawRoleFields.MEMBER.has(field)).length,
    classifiedHeadFields: head.size,
    classifiedMemberFields: member.size,
    classifiedHeadOnlyFields: [...head].filter((field) => !member.has(field)).length,
    sixMonthFields: periodic("SIX_MONTHS").size,
    sixMonthHeadFields: periodicByRole("SIX_MONTHS", "HEAD").size,
    sixMonthMemberFields: periodicByRole("SIX_MONTHS", "MEMBER").size,
    annualFields: periodic("ANNUAL").size,
    annualHeadFields: periodicByRole("ANNUAL", "HEAD").size,
    annualMemberFields: periodicByRole("ANNUAL", "MEMBER").size,
    skippedUnknownHeaders: [...skippedUnknown].sort(),
  };
  const expected = {
    totalRows: 336,
    activeRows: 294,
    inactiveRows: 42,
    classifiedSchemaFields: 229,
    headFields: 224,
    memberFields: 66,
    headOnlyFields: 164,
    memberOnlyFields: 6,
    sharedFields: 60,
    classifiedHeadFields: 223,
    classifiedMemberFields: 66,
    classifiedHeadOnlyFields: 163,
    sixMonthFields: 81,
    sixMonthHeadFields: 78,
    sixMonthMemberFields: 32,
    annualFields: 26,
    annualHeadFields: 26,
    annualMemberFields: 3,
    skippedUnknownHeaders: ["alamat_ktp"],
  };
  if (JSON.stringify(stats) !== JSON.stringify(expected)) {
    console.error(JSON.stringify(stats, null, 2));
    throw new Error("Generated metadata stats differ from the reviewed CSV baseline.");
  }

  const output = {
    _meta: {
      sources: ["config/sources/merge_pertanyaan.csv"],
      tooltipPolicy: previous._meta?.tooltipPolicy ?? "QUESTION_LOGIC_ONLY",
      unmappedFrequency: Object.keys(mapping).filter((field) => !fields[field].frequency),
      stats,
    },
    fields,
    events: previous.events ?? {},
    building: previous.building ?? {},
  };
  fs.writeFileSync(METADATA_PATH, `${JSON.stringify(output, null, 2)}\n`);
  console.log(JSON.stringify(stats, null, 2));
}

try {
  main();
} catch (error) {
  console.error(error instanceof Error ? error.message : error);
  process.exitCode = 1;
}
